### PR TITLE
Feature batch add function

### LINKED ISSUE #79: Add Batch Add Function
As a new user, I can batch-add people into the database, so that I can efficiently transfer my current employee list into programme.

The batch-add function should allow users to "upload" or add multiple employee's data at once.
This should be done through:

1. Reading of data from a `csv` file
2. Processing of rows from the `csv` file into `Person` objects
3. Return the appropriate model and save as `json`

At any point in time, if there are erroneous entries, the process should throw an exception and terminate. It would be good to highlight the location of the error as well.

### PR DESCRIPTION
**Pull request by prit3010**
Fixes #79 

Added Batch Add Functionality where the User can add a group of people at once instead of adding them one by one.

Brief Overview on the feature:
- CSV File containing all the information is stored in `data` folder
- Type the Command Prompt with the filename into the system
- `BatchAddParser` will parse for the filename
- `getInfo()` reads the csv file and returns a `List<AddCommand>`
- When `execute` is run, will execute all the AddCommands in the List

Additional Features:
- If duplicates are found in the list of the CSV, the whole function will be aborted

Test Cases:
- Test Cases are still limited
- Unsure on how to create Test case for success

### CODE REVIEW

File: src/main/java/coydir/logic/commands/BatchAddCommand.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+                            arg += PREFIX_LIST[i] + tag + " ";
+                        }
+                    } else {
+                        arg += PREFIX_LIST[i] + data[i] + " ";
+                    }
+                }
+                addCommandList.add(new AddCommandParser().parse(arg));
+            }
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
```
Comment (kevinchangjk) [inline src/main/java/coydir/logic/commands/BatchAddCommand.java:74]: Applies for line 77 below also.

Can we not print the error stack? It kinda ruins the logging in the terminal.

image

**Review by kevinchangjk — CHANGES_REQUESTED** [review]
There's a couple things that needs changing.

image

Here, I ran `batchadd BatchAddSuccess.csv` once, deleted the two new entries, then ran it again. The numbers are a bit funny cause I did more stuff before this sequence, but the point here is that the `EmployeeId` fields are not reverted when a batch add command fails.

Might have to call the `EmployeeId::setCount` method after failing, to reset it to before the command.

Additionally, you can see the other comments I left:
- some minor coding changes/suggestions
- extra blank lines
- `System.out.println` left behind

File: src/main/java/coydir/logic/parser/AddCommandParser.java [hunk 2]
```diff
@@ -34,6 +34,7 @@ public class AddCommandParser implements Parser<AddCommand> {
      * @throws ParseException if the user input does not conform the expected format
      */
     public AddCommand parse(String args) throws ParseException {
+        System.out.println(args);
```
[the remark this answers is not in the document]
Comment (prit3010) [reply src/main/java/coydir/logic/parser/AddCommandParser.java:37]: Resolved

File: src/main/java/coydir/logic/parser/ParserUtil.java [hunk 3]
```diff
@@ -125,6 +125,7 @@ public static Position parsePosition(String position) throws ParseException {
         return new Position(trimmedPosition);
     }
 
+
```
[the remark this answers is not in the document]
Comment (prit3010) [reply src/main/java/coydir/logic/parser/ParserUtil.java:128]: Resolved

File: src/main/java/coydir/model/ModelManager.java [hunk 4]
```diff
@@ -13,6 +13,7 @@
 import javafx.collections.ObservableList;
 import javafx.collections.transformation.FilteredList;
 
+
```
[the remark this answers is not in the document]
Comment (prit3010) [reply src/main/java/coydir/model/ModelManager.java:16]: Resolved

File: src/main/java/coydir/model/ModelManager.java [hunk 5]
```diff
@@ -104,6 +105,7 @@ public void addPerson(Person person) {
         updateFilteredPersonList(PREDICATE_SHOW_ALL_PERSONS);
     }
 
+
```
[the remark this answers is not in the document]
Comment (prit3010) [reply src/main/java/coydir/model/ModelManager.java:108]: Resolved

File: src/test/java/coydir/logic/commands/AddCommandTest.java [hunk 6]
```diff
@@ -34,9 +34,10 @@ public void constructor_nullPerson_throwsNullPointerException() {
     public void execute_personAcceptedByModel_addSuccessful() throws Exception {
         ModelStubAcceptingPersonAdded modelStub = new ModelStubAcceptingPersonAdded();
         Person validPerson = new PersonBuilder().build();
+        System.out.println(validPerson.toString());
 
         CommandResult commandResult = new AddCommand(validPerson).execute(modelStub);
-
+        System.out.println(modelStub.personsAdded.get(0).toString());
```
[the remark this answers is not in the document]
Comment (prit3010) [reply src/test/java/coydir/logic/commands/AddCommandTest.java:40]: Resolved

File: src/test/java/coydir/logic/commands/AddCommandTest.java [hunk 7]
```diff
@@ -113,6 +116,7 @@ public void addPerson(Person person) {
             throw new AssertionError("This method should not be called.");
         }
 
+
```
[the remark this answers is not in the document]
Comment (prit3010) [reply src/test/java/coydir/logic/commands/AddCommandTest.java:119]: Resolved

File: src/test/java/coydir/testutil/TypicalPersons.java [hunk 8]
```diff
@@ -92,4 +92,9 @@ public static Database getTypicalDatabase() {
     public static List<Person> getTypicalPersons() {
         return new ArrayList<>(Arrays.asList(ALICE, BENSON, CARL, DANIEL, ELLE, FIONA, GEORGE));
     }
+
+    public static List<Person> getRestPersons() {
+
```
[the remark this answers is not in the document]
Comment (prit3010) [reply src/test/java/coydir/testutil/TypicalPersons.java:97]: Resolved

File: src/main/java/coydir/logic/parser/BatchAddCommandParser.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    public BatchAddCommand parse(String args) throws ParseException {
+        try {
+            String filename = ParserUtil.parseFileName(args);
+            return new BatchAddCommand(filename);
+        } catch (ParseException pe) {
+            throw new ParseException(
+                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, BatchAddCommand.MESSAGE_USAGE), pe);
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (prit3010) [reply src/main/java/coydir/logic/parser/BatchAddCommandParser.java:28]: Resolved

File: src/main/java/coydir/logic/commands/BatchAddCommand.java [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+        if (addCommandList.isEmpty()) {
+            throw new CommandException(String.format("%s does not have any data", this.filename));
+        }
+        try {
+            for (AddCommand item : addCommandList) {
+                item.execute(model);
+            }
+        } catch (CommandException e) {
+            Database ab = new Database();
```
[the remark this answers is not in the document]
Comment (prit3010) [reply src/main/java/coydir/logic/commands/BatchAddCommand.java:100]: `model.getDatabase()` returns a `ReadOnlyDatabase` which does not have the `setPersons` method. So I have to instantiate a new `Database` to write into it

File: src/main/java/coydir/logic/commands/BatchAddCommand.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        List<AddCommand> addCommandList = this.getInfo();
+        List<Person> copyOfPersonList = new ArrayList<>();
+        for (Person p : model.getDatabase().getPersonList()) {
+            copyOfPersonList.add(p);
+        }
+
+        if (addCommandList.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (prit3010) [reply src/main/java/coydir/logic/commands/BatchAddCommand.java:92]: Resolved

File: src/main/java/coydir/logic/commands/AddCommand.java [hunk 12]
```diff
@@ -38,7 +38,7 @@ public class AddCommand extends Command {
             + PREFIX_TAG + "owesMoney";
 
     public static final String MESSAGE_SUCCESS = "New person added: %1$s";
-    public static final String MESSAGE_DUPLICATE_PERSON = "This person already exists in the database";
+    public static final String MESSAGE_DUPLICATE_PERSON = "%1$s already exists in the address book";
```
[the remark this answers is not in the document]
Comment (prit3010) [reply src/main/java/coydir/logic/commands/AddCommand.java:41]: Resolved

**Review by kevinchangjk — APPROVED** [review]

**Pull request merged**